Add Onboarding Wiz. Purpose step

yscik opened the pull request:
Fixes #3039 

### Changes proposed in this Pull Request

* Add Purpose step with fixed options checkbox list
* Also contains some CSS improvements, and greener controls
* REST API (saving things) is not yet hooked up

### Testing instructions

* Navigate to `/wp-admin/admin.php?page=sensei_onboarding&step=purpose`
* Items should be selectable
* Text field appears when 'Other' is selected
* Continue navigates to next page

### Screenshot

<img width="1092" alt="Sensei Onboarding Purpose step" src="https://user-images.githubusercontent.com/176949/81015443-03805200-8e5f-11ea-8a2b-5e8177609bd3.png">

donnapep commented:
Looking great!

A couple of small things. The checkbox is slightly misaligned with the label (it's a bit too high):

![Screen Shot 2020-05-06 at 8 53 28 AM](https://user-images.githubusercontent.com/1190420/81179112-374e9b00-8f77-11ea-873a-3ca04b3a4800.jpg)

Could we set placeholder text on the Other field?
![Screen Shot 2020-05-06 at 8 55 39 AM](https://user-images.githubusercontent.com/1190420/81179249-649b4900-8f77-11ea-97cd-2ef6c97bd2c3.jpg)

If I select _Other_ I can't continue until I enter something in the box, but I can continue if I select some other checkbox in addition to _Other_. Given this, I'm wondering if it might be better to not make the text field required. Hopefully most people will fill it out, but I don't want to slow anyone down if they'd rather not.

renatho commented on assets/onboarding/purpose/index.jsx:
```diff
@@ -0,0 +1,120 @@
+							className="sensei-onboarding__textcontrol-other"
+							value={ other }
+							onChange={ ( value ) =>
+								setState( { selected, other: value } )
```
I think in this case would be a better practice to use the `setState` with callback to get the previous `selected` value. Something like `setState( ( { selected } ) => ( { selected, other: value } ) )`

yscik replied:
Updated in aff9621 to merge changes with previous state.

yscik commented:
> A couple of small things. The checkbox is slightly misaligned with the label (it's a bit too high):

Fixed in 5b62a53.

> 
> Could we set placeholder text on the Other field?

Sure, cbbc6d7 

> If I select _Other_ I can't continue until I enter something in the box, but I can continue if I select some other checkbox in addition to _Other_. Given this, I'm wondering if it might be better to not make the text field required. Hopefully most people will fill it out, but I don't want to slow anyone down if they'd rather not.

Yes, I was checking for something to be given: either a thing selected or some text if only Other is selected. Removed the latter in 78221d5.

yscik commented on assets/onboarding/purpose/index.jsx:
```diff
@@ -0,0 +1,120 @@
+export const Purpose = () => {
+	const { goTo } = useQueryStringRouter();
+
+	const [ { selected, other }, setState ] = useState( {
```
Went with `formState`, `setFormState` as it also includes the 'Other' text field value.

renatho replied:
Makes sense! =)

yscik commented on assets/onboarding/purpose/index.jsx:
```diff
@@ -0,0 +1,120 @@
+		! selected.length ||
+		( isEqual( [ 'other' ], selected ) && ! other.length );
+
+	function selectItem( id ) {
```
Starting to regret with went with that :) 
efc840e

renatho replied:
Don't you like it?

yscik commented on assets/onboarding/purpose/index.jsx:
```diff
@@ -0,0 +1,121 @@
+
+	const isEmpty = ! selected.length;
+
+	const selectItem = ( id ) => {
```
That's better, yes, updated in a21a77e.
Also made one more minor change, putting the Other option in the purpose list as it doesn't require any special handling.

yscik commented:
Rebased onto master.

Also made one more minor change in f10624a, putting the Other option in the purpose list as it doesn't require any special handling.
